### ISSUE TITLE
New-AzureRmResourceGroupDeployment with -TemplateParameterObject doesn't support parameter named "name"

### ISSUE THREAD

**Issue by gregjhogan** [Azure/azure-powershell#2278]
ARM template with parameter named `name`

``` ps1
$template = @{
    '$schema' = "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#"
    contentVersion = "1.0.0.0"
    parameters = @{
        name = @{ type = "string" }
    }
    variables = @{}
    resources = @()
    outputs = @{}
}
ConvertTo-Json -InputObject $template -Depth 1024 | Out-File -File template.json -Force
$templateParams = @{ name = "name-test" }
New-AzureRmResourceGroupDeployment -ResourceGroupName test-rg -TemplateFile template.json -TemplateParameterObject $templateParams
```

Causes a prompt for a parameter `nameFromTemplate`

```
cmdlet New-AzureRmResourceGroupDeployment at command pipeline position 1
Supply values for the following parameters:
(Type !? for Help.)
nameFromTemplate:
```

If you give it an object representing all the template parameters, it seems like it shouldn't be getting confused.

**Comment by ravbhatnagar** [comment]
This is currently by design. If a parameter name used in the template conflicts with the name of an inline parameter (Name is a valid parameter for this cmdlet), then we append the keyword fromTemplate to distinguish between the two.
Please let me know if this sounds reasonable and I will close the issue.

**Comment by gregjhogan** [comment]
@ravbhatnagar but why is that necessary in my situation?

What you are saying makes sense when passing the template parameters as additional cmdlet parameters
```ps1
New-AzureRmResourceGroupDeployment -ResourceGroupName 'test-rg' -TemplateFile 'template.json' -Name 'deployment-name' -NameFromTemplate 'name-test'
```

But why should there be any confusion when it is getting passed through `-TemplateParameterObject`?
```ps1
New-AzureRmResourceGroupDeployment -ResourceGroupName 'test-rg' -TemplateFile 'template.json' -Name 'deployment-name' -TemplateParameterObject @{ name = "name-test" }
```

The current design is functionally complete, but it doesn't seem ideal because it was pretty confusing initially when it started prompting me for `nameForTemplate`.  I would argue it is confusing because it is unexpected when using `-TemplateParameterObject` and in my specific scenario it is especially confusing because prompting for `nameForTemplate` can be interpreted as asking for a template name (as opposed to the name parameter in the template).

**Comment by ravbhatnagar** [comment]
I agree, it can be confusing. Let me see if we can solve this for TemplateParameterObject.

**Comment by gregjhogan** [comment]
@ravbhatnagar looks like there is actually a bug.  I can't find a way to use `-TemplateParameterObject` and pass a parameter named `name`.

The following doesn't prompt for any parameters
```ps1
$template = @{
    '$schema' = "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#"
    contentVersion = "1.0.0.0"
    parameters = @{
        name = @{ type = "string" }
    }
    variables = @{}
    resources = @()
    outputs = @{}
}
ConvertTo-Json -InputObject $template -Depth 1024 | Out-File -File template.json -Force
$templateParams = @{ nameFromTemplate = "name-test" }
New-AzureRmResourceGroupDeployment -ResourceGroupName test-rg -TemplateFile template.json -TemplateParameterObject $templateParams
```

and throws an error saying `nameFromTemplate` is not a valid parameter name.
```
WARNING: The output object type of this cmdlet will be modified in a future release.
New-AzureRmResourceGroupDeployment : 4:23:48 PM - Error: Code=InvalidTemplate; Message=Deployment
template validation failed: 'The template parameters 'nameFromTemplate' in the parameters file are
not valid; they are not present in the original template and can therefore not be provided at
deployment time. The only supported parameters for this template are 'name'. Please see
http://aka.ms/arm-deploy/#parameter-file for usage details.'.
At line:13 char:1
+ New-AzureRmResourceGroupDeployment -ResourceGroupName test-group -Tem ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [New-AzureRmResourceGroupDeployment], Exception
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.NewAzureResourceGroupDeploymentCommand

New-AzureRmResourceGroupDeployment : The deployment validation failed.
At line:13 char:1
+ New-AzureRmResourceGroupDeployment -ResourceGroupName test-group -Tem ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [New-AzureRmResourceGroupDeployment], InvalidOperationException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.NewAzureResourceGroupDeploymentCommand
```

**Comment by ravbhatnagar** [comment]
@gregjhogan The current issue is that if you use a parameter name like "name" and specify the value in the parameter Object, it is ignoring that and asking you for the value again. It appends FromTemplate to the param (called name in this case). But in the output, you can see the value you provide for nameFromTemplate is actually used to update the value of "name" parameter. 

In above example you should specify templateParams object as $templateParams = @{ name = "name-test" }

You see the above issue, as the parameter name is not defined in the template file. But we will fix the other issue where it should not prompt you if the param is already present in the object.

**Comment by gregjhogan** [comment]
@ravbhatnagar can you please fix this?!  When using either -TemplateParameterFile or -TemplateParameterObject there is no non-interactive way to pass a parameter named "name".  I would say when it is in the parameter file or parameter object it should know that is what you want to pass for the value.  If that is not possible, please at least support what it asks for when in interactive mode by expecting a parameter named `nameFromTemplate` in my parameter file or object.

**Comment by fleed** [comment]
Any plans to fix this? I'm running into the same issue. How can I make the script run seamlessly in a `Visual Studio Team Services` release?

**Comment by ravbhatnagar** [comment]
@fleed - If you have a parameter name "Name" in your template, it will prompt you to enter value for parameter - "nameFromTemplate". This is because the cmdlet itself has a parameter called Name. So, an easy fix would be to rename the parameter.

We will fix the other issue which is as follows -The current issue is that if you use a parameter name like "name" and specify the value in the parameter Object, it is ignoring that and asking you for the value again. It appends FromTemplate to the param (called name in this case).

**Comment by tinymoth** [comment]
I'm suffering from this when I export templates from Azure. If you export a template/parameter set from a created resource, the 'name' fieldname is extended to match the resource in the template and the corresponding parameters file. If you try to export a resource by filling in the values in the gui and clicking on Automation Options, it just uses the 'name' fieldname for the parameter, thereby rendering the automation broken. 
If you try to use PS to suck in the parameter file with convertfrom-json so that you remove the param, the command fails because the schema does not mark 'name' as valid. There is no -force to accept it, unfortunately.
Editing files produced by MS to be consumed by the MS script provided in the same archive is not desirable. I don't want to have to tell clients that the easy tool has a bunch of manual editing steps to fix the problems it creates.
So, please let name be used, perhaps with a -pleaseUseTheNameInTheParameterFileEvenThoughThisCmdletHasItsOwnNameParameterBecauseImNotConfused flag.
Cheers,
Tim
[broken-template.zip](https://github.com/Azure/azure-powershell/files/586226/broken-template.zip)

**Comment by chrisinazure** [comment]
small note: if you export the templates created by msft through the portal, you have the same issue. workaround at the moment (although silly) is replace the name parameter with for example appname, solved. something that can be easily fixed by msft i'm guessing? ;-)

**Comment by i255d** [comment]
ravbhatnagar, the thing you are not making clear here, is where to put:
$templateParams = @{ name = "name-test" }

This is the template:
"parameters": {
        "name": {
            "type": "string"
        },


This is the parameter file:

   "parameters": {
        "type": {
            "value": "web"
        },
        "name": {
            "value": ""
        },

or does the hash get puting the parameter or piped into the script some how?

**Comment by angeloedades** [comment]
i255d,

"name" needs to be renamed, in your ARM template, with a different name so name it so it does not conflict with a parameter called "name" in the cmdlet.

**Comment by gregjhogan** [comment]
@ravbhatnagar any idea when you will make the fix you previously mentioned?  Right now there is no noninteractive way to deploy a template with a parameter named 'name'.  Thanks!

**Comment by i255d** [comment]
@angeloedades  @ravbhatnagar
I thought my message made it pretty clear that I understand that "name" needs to be renamed, the questions is where?  I don't know what a ARM template is.  (Azure Resource Manager?)

I did get it to work.  First I changed it in the JSON Parameter file:

{
    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "type": {
            "value": "web"
        },
        "name-test": {
            "value": "Message-Marker-Prod"
        },
        "hockeyAppToken": {
            "value": ""
        },
        "hockeyAppId": {
            "value": ""
        },
        "regionId": {
            "value": "southcentralus"
        },
        "requestSource": {
            "value": "IbizaAIExtension"
        }
    }
}

Then I adapted the JSON templet but left the "name" in the resouces:

{
    "$schema": "http://schema.management.azure.com/schemas/2014-04-01-preview/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "name-test": {
            "type": "string"
        },
        "type": {
            "type": "string"
        },
        "hockeyAppToken": {
            "type": "string"
        },
        "hockeyAppId": {
            "type": "string"
        },
        "regionId": {
            "type": "string"
        },
        "requestSource": {
            "type": "string"
        }
    },
    "resources": [
        {
            "name": "[parameters('name-test')]",
            "type": "microsoft.insights/components",
            "location": "[parameters('regionId')]",
            "apiVersion": "2014-08-01",
            "properties": {
                "ApplicationId": "[parameters('name-test')]",
                "Application_Type": "[parameters('type')]",
                "HockeyAppToken": "[parameters('hockeyAppToken')]",
                "HockeyAppId": "[parameters('hockeyAppId')]",
                "Flow_Type": "Redfield",
                "Request_Source": "[parameters('requestSource')]"
            }
        }
    ]
}

Again this seemed to work.
Everywhere it says Name-Test has been changed from Name

**Comment by cormacpayne** [comment]
@ravbhatnagar any update on this issue?

**Comment by hobpet** [comment]
It would be a good enough solution to check this during the template validation and provide a meaningful error message.

**Comment by jlarrow** [comment]
I've also encountered this bug.  Not sure why data **inside** a parameter of type Hashtable would conflict with a named parameter.  Is this because "-TemplateParameterObject" has ValueFromPipelineByPropertyName on it?   In the mean time, I guess I'll rename my parameter.  I'd be interested in having a deeper understanding though.

**Comment by abatishchev** [comment]
What worked for me: don't have a parameters called "name", this will cause the prompt. Use something else, e.g. resourceName.

**Comment by holgerjay** [comment]
Agreed, @abatishchev , that is an easy fix. It is a bit strange though that the conflict is introduced when downloading the pre-created template from Azure.

**Comment by BrainSlugs83** [comment]
It also fails (prompting for this 'name' parameter) when you pass it in by file -- which is very lame, considering some of the default Azure RM templates are broken (i.e. if you go to create a new Storage account and export the template there, it doesn't work -- it makes you pass in the `nameFromTemplate` parameter, even though it's stored right there in the .json file).

**Comment by seanfarmar** [comment]
Is this going to be fixed? This is a real bummer, in order for this to work I need to modify all the json files generated for each deployment...

**Comment by Chanticrow** [comment]
I'm a new Azure user trying to implement templates based on the exported files from the portal.  Would someone from Microsoft at least go update the exported templates so that they don't use "name" as a parameter?  I'm grateful that I found this two and a half year old thread that explained what I needed to fix.

**Comment by gmoed** [comment]
I experienced the same issue (parameter name apiVersion; this was used in the parameter object passed to the template 

![image](https://user-images.githubusercontent.com/28516433/49847539-c740f180-fe24-11e8-8296-91f9a531fabc.png)

and its also a valid parameter name in the cmdlet 'New-AzureRmResourceGroupDeployment' ), solved it by changing the parameter name in the cmdlet

ex: 
![image](https://user-images.githubusercontent.com/28516433/49847510-b2645e00-fe24-11e8-91ce-6c813c95c45a.png)

**Comment by abatishchev** [comment]
@ravbhatnagar hi, any plans to address this issue anyhow? It's a source of troubles and frustration for so many customers for several years now.

**Comment by LukeCarrier** [comment]
Please stop +1ing in the thread and just click the :+1: button in the issue
description. Everyone subsribed to this issue receives emails when you
reply to it.

**Comment by abatishchev** [comment]
That's the exactly the point of replying, it bumps the thread. Expressing an emotion does not.

**Comment by LukeCarrier** [comment]
I think Microsoft have demonstrated their indifference pretty clearly in
this thread -- there has been no progress on this issue since it was
raised. I share your frustrating, but I don't think carpet bombing all of
the issue's subscribers with emails is helping anyone.

</carpet bombing all of the issue's subscribers>

**Comment by gmoed** [comment]
@abatishchev , I provided a way to overcome this "issue" in a comment above. rename the parameter passed in the cmdlet as shown in the pic, that should work for any conflicting parameter name

**Comment by abatishchev** [comment]
@gmoed I know to work this around. Rename every custom parameter of which name is reserved. But figure this out from weird, poorly worded, undocumented behavior. Right.

Instead, I proposed a better validation with better message explaining what's wrong and what user should do about it.

**Comment by englishbobsnr** [comment]
Hey hey hey who's here in 2019! My first real step into IaC and I spent a day on trying to work out what I was doing wrong. Turns out I wasn't doing anything wrong and this is a bug from 2016 !! You have to be joking me. @tinymoth hit the nail on the head in 2016, yet 3 years later, nothing

**Comment by ViswanathAkondi** [comment]
Agreed i have spent many days trying to figure out if it was missing mentioning any parameters in the file 
and it turns out that this is a Bug . 
I have been trying to create a unique name for storage account everytime i run the template using the Guid feature and everytime it would be needing me to specify the name and template.json to get the resource template working. However post giving these details it works absolutely fine . Please assist in fixing this bug .

**Comment by aperisic** [comment]
I cannot believe that I am reading this. You export a template and when you try to run it, it gives you something is bad. You read and find that "it is as it is" and that you have to patch it yourself.

The issue is very simple for me. Azure is supposed to give a developer power to deploy effortlessly. This is a classical operational task of fixing a script for whatever reason.

I am not even worried that this particular problem exists, I am worried about mentality that does not see how problems like this one diverge from the advertised generality of Azure. If I am to go around fixing things like this one, then it is all the same if I create an ad-hoc solution myself as It will have the same number of problems.

Things like this are telling me that instead of creating a one-step solution I will every now and then spend hours of figuring out how Azure really works reading blogs about undocumented behaviors.

If I would buy a product with that kind of flaws, I would not keep buying it.

**Comment by alex-frankel** [comment]
I sincerely apologize for this longstanding issue not being addressed for such a long time. We have a PR out with the fix, so this issue should be gone in the next release of the Microsoft.Resources module. FWIW, this issue does not eseem to exist in Az CLI if you would prefer to go that route.

cc @shenglol

**Comment by angeloedades** [comment]
YAY! @alex-frankel

**Comment by rthorn17** [comment]
@alex-frankel - Is this item closed now?

**Comment by aperisic** [comment]
Closed merged fixed done and completed

**Comment by shenglol** [comment]
Already fixed in Az 3.2.0. Closing the issue.
